Tweaks to invalid ctor messages

estebank opened the pull request:
- Do not suggest using a constructor that isn't accessible
 - Suggest the appropriate syntax (`()`/`{}` as appropriate)
 - Add note when trying to use `Self` as a ctor

CC #22488, fix #47085.

estebank committed:
Hide suggestion to use struct ctor when it is not visible

estebank committed:
Suggest the correct syntax for different struct types

estebank committed:
Readd suggestion in enum variants with incorrect args

rust-highfive commented:
r? @michaelwoerister

(rust_highfive has picked a reviewer for you, use r? to override)

estebank committed:
Add note when trying to use `Self` as a ctor

estebank commented on src/test/ui/resolve/tuple-struct-alias.stderr:
```diff
@@ -2,31 +2,29 @@ error[E0423]: expected function, found self type `Self`
-   |                 ^^^^ did you mean `Self { /* fields */ }`?
+   |                 ^^^^ not a function
+   |
+   = note: can't instantiate `Self`, you must use the implemented struct directly
```
This message would be better if it mentioned the struct `S`. It'd be even better if it checked that it was a suggestion:

```
  = help: can't instantiate `Self`, you must use the struct `S` directly:
   |
16 |         let s = S(0, 1); //~ ERROR expected function
   |                 ^

petrochenkov commented:
@estebank 
Sorry for the delay, I'll review all the PRs in a couple of days.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
+                                                     block));
+                        return (err, candidates);
+                    }
+                    (Def::SelfTy(_, _), _) if ns == ValueNS && is_struct_like(def) => {
```
`is_struct_like` is always `true` here.

petrochenkov replied:
Nit: `_, _` -> `..`

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
+                        return (err, candidates);
+                    }
+                    (Def::SelfTy(_, _), _) if ns == ValueNS && is_struct_like(def) => {
+                        err.note("can't instantiate `Self`, you must use the implemented struct \
```
I don't think "instantiate" is a correct word here.
Maybe "can't use `Self` as a constructor"?

petrochenkov replied:
Also the suggestion to use `Self { ... }` with braces is now missing (but maybe it's okay).

petrochenkov replied:
Also suggestions to use `Foo { ... }` with braces are now missing for type aliases.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
-                                    err.span_label(span, format!("constructor is not visible \
-                                                                   here due to private fields"));
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS && is_struct_like(def) => {
```
`is_struct_like` is always `true` here.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
-                                                     path_str));
                         return (err, candidates);
                     }
+                    (Def::VariantCtor(_, ctor_kind), _) if ns == ValueNS && is_struct_like(def) => {
```
This is equivalent to `(Def::VariantCtor(_, CtorKind::Fictive), _) if ns == ValueNS`

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
+                                    Def::StructCtor(_, CtorKind::Fn) |
+                                    Def::VariantCtor(_, CtorKind::Fn) => "(/* fields */)",
+                                    Def::StructCtor(_, CtorKind::Fictive) |
+                                    Def::VariantCtor(_, CtorKind::Fictive) => {
```
Variants ctors are unreachable here and fictive struct ctor is impossible.

petrochenkov replied:
Could you also add a test for this suggestion?

estebank replied:
So I'm is it correct for me to understand that in this case I should be checking for `ctor_def`` being `Def::Struct(..)` instead of `Def::StructCtor(_, Fictive)`?

estebank committed:
address review comments

estebank committed:
Further tweaks to the output

- Properly address Variant Ctors
- Show signature if span of trait method without `self` is not available

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,20 +2702,51 @@ impl<'a> Resolver<'a> {
+                            if is_expected(ctor_def) && !accessible_ctor {
+                                err.span_label(span, format!("constructor is not visible \
+                                                              here due to private fields"));
+                            } else if accessible_ctor {
```
Actually, now I don't think this condition is reachable at all.
With `ns == Value` we would find the constructor before `Def::Struct` if it were really accessible.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,20 +2702,51 @@ impl<'a> Resolver<'a> {
+                                                      path_str));
+                        return (err, candidates);
+                    }
+                    (Def::VariantCtor(_, CtorKind::Fictive), _) => {
```
These three `VariantCtor` arms look like they give tautological recommendations, I don't think they are ever useful (at least in realistic code).

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,20 +2702,51 @@ impl<'a> Resolver<'a> {
                     }
+                    (Def::SelfTy(..), _) if ns == ValueNS => {
+                        err.note("can't use `Self` as a constructor, you must use the \
+                                  implemented struct");
```
Missing `return (err, candidates);`

estebank replied:
Added.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,20 +2702,51 @@ impl<'a> Resolver<'a> {
+                                  implemented struct");
+                    }
+                    (Def::TyAlias(_def_id), _) => {
+                        err.note("can't use a type alias as a constructor");
```
Missing `return (err, candidates);`, missing `if ns == ValueNS`, unused `_def_id`.

estebank replied:
Added.

estebank committed:
Review comments: remove enum suggestion

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 282f75a has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 282f75a4ced2b859b10779ea91d7e889ddfdc659 -->

estebank commented:
@bors r-, there're some diagnostic output I need to fix by re-adding `VariantCtor` match arms. Is it ok if I "r" it after doing that or should I change the output to match the current code?

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,18 +2702,27 @@ impl<'a> Resolver<'a> {
+                                  implemented struct");
+                        return (err, candidates);
+                    }
+                    (Def::TyAlias(_), _) if ns == ValueNS => {
```
`Def::AssociatedTy(..)` can be added here.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -2700,18 +2702,27 @@ impl<'a> Resolver<'a> {
-                                    err.span_label(span, format!("constructor is not visible \
-                                                                   here due to private fields"));
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS => {
```
`Def::Union(..)`, `Def::Variant(..)` and `Def::VariantCtor(_, CtorKind::Fictive)` can be added here to restore what the previous code did (they are equivalent to structs without accessible constructors in this context).

petrochenkov replied:
Only `CtorKind::Fictive` is relevant here because for others https://github.com/rust-lang/rust/pull/47116#discussion_r161618498 still holds.

petrochenkov commented:
Some other cases beside `VariantCtor` are missing too.

kennytm commented:
@bors r-

The command in https://github.com/rust-lang/rust/pull/47116#issuecomment-358117745 is ignored since homu treats `r-,` as a single word 😒

estebank committed:
Fix tests by keepeing needed suggestions

petrochenkov commented:
@bors r+

bors commented:
:hourglass: Testing commit 067405044cb682f1d01c8665e81287c23dd3655f with merge 97520ccb101609af63f29919bb0a39115269c89e...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6096), [status-travis](https://travis-ci.org/rust-lang/rust/builds/331501712?utm_source=github_status&utm_medium=notification)
Approved by: petrochenkov
Pushing 97520ccb101609af63f29919bb0a39115269c89e to master...
